File: examples/sogou_weixin_search.py
#!/usr/bin/env python3
"""
SougouWeixin search example for the tarzi Python library.
Demonstrates searching WeChat articles using Sogou's WeChat search engine.
"""

import tarzi


def main():
    """Main function to demonstrate SougouWeixin search functionality."""
    print("=== Tarzi Python SougouWeixin Search Example ===\n")

    # Create configuration for SougouWeixin search
    config_str = """
[search]
engine = "sogou_weixin"
limit = 10

[fetcher]
web_driver = "chromedriver"
mode = "browser_head"
timeout = 30
format = "markdown"
"""
    
    try:
        config = tarzi.Config.from_str(config_str)
    except Exception as e:
        print(f"Failed to create config: {e}")
        # Fallback to default config
        config = tarzi.Config()
        print("Using default configuration")

    # Create search engine from config
    search_engine = tarzi.SearchEngine.from_config(config)

    # Query about Oracle Corporation stock price in Chinese
    query = "甲骨文股价"
    print(f"\nSearching WeChat articles for: '{query}'")
    # Perform the search
    try:
        results = search_engine.search(query, 10)
        print(f"\nFound {len(results)} WeChat articles:")
        if not results:
            print("No results found.")
        else:
            for i, result in enumerate(results):
                print(f"\n{i + 1}. {result.title}")
                print(f"   URL: {result.url}")
                if result.snippet:
                    print(f"   Snippet: {result.snippet}")
                print(f"   Rank: {result.rank}")

            print("\n=== Search Summary ===")
            print(f"Total results: {len(results)}")
            print("All results are from mp.weixin.qq.com (WeChat articles)")

    except Exception as e:
        print(f"Search failed: {e}")
        raise

    # Ensure clean shutdown of browser and driver resources
    try:
        search_engine.cleanup()
    except Exception as e:
        print(f"Warning: Failed to properly cleanup search engine: {e}")

    print("\n=== SougouWeixin Search Example completed successfully! ===")


if __name__ == "__main__":
    main()

File: examples/sogou_weixin_search.rs
use std::str::FromStr;
use tarzi::{
    Result,
    config::Config,
    search::{SearchEngine, types::SearchEngineType},
};

#[tokio::main]
async fn main() -> Result<()> {
    // Initialize logging
    tracing_subscriber::fmt::init();

    let mut config = Config::load()?;
    config.search.engine = "sogou_weixin".to_string();
    config.search.limit = 10;
    let engine_type = SearchEngineType::from_str(&config.search.engine).unwrap();
    config.search.query_pattern = engine_type.get_query_pattern();
    config.fetcher.web_driver = "chromedriver".to_string();
    config.fetcher.mode = "browser_head".to_string();

    // Create search engine from config
    let mut search_engine = SearchEngine::from_config(&config);

    // Query about Oracle Corporation stock price in Chinese
    let query = "甲骨文股价";
    match search_engine.search(query, config.search.limit).await {
        Ok(results) => {
            if results.is_empty() {
                println!("No results found.");
            } else {
                for (i, result) in results.iter().enumerate() {
                    println!("\n{}. {}", i + 1, result.title);
                    println!("   URL: {}", result.url);
                    if !result.snippet.is_empty() {
                        println!("   Snippet: {}", result.snippet);
                    }
                    println!("   Rank: {}", result.rank);
                }
            }
        }
        Err(e) => {
            eprintln!("Search failed: {e}");
            return Err(e);
        }
    }

    // Ensure clean shutdown of browser and driver resources
    search_engine.shutdown().await;

    Ok(())
}
